wordpress-mobile/WordPress-FluxC-Android
# Woo: Update product fields

Fixes #1435. This PR adds logic to update the following product fields to the API.

#### Changes
- Adds two columns/fields `dateOnSaleFrom` and `dateOnSaleTo` to `WCProductModel` to allow for scheduling product sales.
- Adds logic to update product only if updated product fields do not match the stored product fields. This is to allow the option for users to update empty product descriptions (should they choose to do so) and also, to ensure that we do not update product fields that do not contain any changes. 
- Testing the API, I found that we cannot update an existing SKU value from another product to the product we are trying to update. So added a new error type in `ProductErrorType` if user tries to update the SKU of the product by entering an existing value.
- Added enum classes for displaying product tax status options (`CoreProductTaxStatus`), product back order options (`CoreProductBackOrders`) and product stock status options (`CoreProductStockStatus`).
- Added a new page in the example app to test updating multiple product fields.

#### Testing
- Run tests in `MockedStack_WCProductsTest`, `WCProductStoreTest`.
- Since there is a migration script added to the PR, it would be good to verify if app can be updated to this PR without any issues.
- Testing in the example app:
  - Open the app, click on `Woo` (after login, of course) -> `Products` and select a site.
  - Click on `Update Product` button.
  - Enter a product ID.
  - Modify fields displayed in the UI and verify that only the fields updated in the UI are updated to the product.
  - Verify that an error is displayed in the example app when trying to update an existing SKU (from another product).

#### Screenshots
<img width="300" src="https://user-images.githubusercontent.com/22608780/70407782-a1682000-1a6b-11ea-8a91-889185db078f.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/70407785-a3ca7a00-1a6b-11ea-87f0-56331a3d612d.png">

#### Notes 
- There is no option to test updating `tax_class` and `shipping_class` fields to a product. This is because we need to add a separate API endpoints to FluxC before testing this. Added issue [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/1442) and here to address this separately.
